rustup-dist: Use Download notifications to track install

kinnison opened the pull request:
People have requested some indication of progress for long-running
install steps.  This commit re-uses the download tracker logic to
provide a progress bar (the length is the compressed component
tarball but should be good enough) to provide such feedback.

Fixes: #1557

kinnison committed:
rustup-dist: Use Download notifications to track install

People have requested some indication of progress for long-running
install steps.  This commit uses the new FileReaderWithProgress to
provide a progress bar (the length is the compressed component
tarball but should be good enough) to provide such feedback.

Fixes: #1557

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

kinnison commented:
All the rebase does is remove one of the fixes messages because it was over-enthusiastic.

nrc reviewed (commented):
Thanks for the PR, I think this is probably a good idea. I would move the reader from the manifestation module to somewhere else though, maybe the dist::download module or a utils module somewhere

kinnison commented:
I will look at the right place to move it to, and attempt to get to it later when I'm home for the evening.  I worry that the callback nesting might take me a while to get right though, so it might be a day or two before I can post depending on how awake I am when I get home.  Thanks for the confirmation that the idea itself is okay though.

kinnison committed:
rustup_utils: Add a FileReaderWithProgress struct

In order to be able to report unpack progress, add support for a file reader
which emits notifications akin to the downloading of a file.  This allows the
generic progress bar supporting downloads to also handle the installation of
components.

Signed-off-by: Daniel Silverstone <dsilvers@digital-scurf.org>

kinnison commented:
@nrc turns out that the FileReaderWithProgress moved nicely to `rustup_utils` but I'm not entirely clear on how I might add tests for this.  Obviously I've run it by hand and am confident it does what I wanted, but do you feel it needs something under `cargo test` ?

kinnison commented:
Interestingly there's some formatting differences in the CI.  I'll have to double-check my rustfmt install.

kinnison commented:
I've rebased and hopefully that'll clear up that formatting issue.

nrc commented:
Thanks!
